### ISSUE TITLE
TabNavigator inside StackNavigator

### ISSUE THREAD

**Issue by puzpuz** [react-community/react-navigation#693]
I saw there is an example to have StackNavigator inside TabNavigator, is it possible to do the other way arround. The reason for this is, because I don't want to show the TabBar in every pages
e.g:
4 tabs: Profile, News, Activities, Settings
Profile Tab
 -- List of friends
 -----Detail of friend

As you see above, List of friend page will be under profile tab, and I dont want to show tab bar on it. As well as Detail of friend page.

So my idea is to stack the TabNavigator inside a StackNavigator, is it possible?

**Comment by matthamil** [comment]
Yep. You should be able to nest it. Are you having any issues doing this?

**Comment by puzpuz** [comment]
This is the first time I learn react-navigation, and I found it difficult to get `this.props.navigation` from the other component (e.g child of a list view).
From example above, I can't get into Detail of friend, when a button on friend list item is pressed. Can you help me understand this? thank you

**Comment by matthamil** [comment]
@puzpuz Can you post the code in question and the expected behavior and actual behavior?

**Comment by puzpuz** [comment]
Stack Navigator that I have is:
```
const FriendListStack = StackNavigator({
  FriendListScreen: {
    screen: FriendListScreen,
  },
  FriendDetailScreen: {
    screen: FriendDetailScreen,
  },
},
{
  initialRouteName: 'FriendListScreen',
  headerMode: 'none'
});
```
inside `FriendListScreen` is a List View with items like this:

```
  renderRow(friend) {
    return (
      <FriendItem 
        friend={friend} 
      />
    );
  }
  render() {
    return (
          <List 
            dataArray={friendData}
            renderRow={this.renderRow}
          />
    );
  }
```
inside `FriendItem` I have a button, and when I click the button I want to redirect the screen to `FriendDetailScreen`

 ```
 onDetailPress() {
    this.props.navigation.navigate('FriendDetailScreen');
  }
  render() {
    return (
      <View>
              <TouchableOpacity
                  onPress={this.onDetailPress.bind(this)}              
              >
                 <Text>Detail</Text>
              </TouchableOpacity>
      </View>
    );
  }
```

How can I pass `this.props.navigation` property from `FriendListScreen` into its child component `FriendItem` ?

**Comment by ronal2do** [comment]
+1

**Comment by matthamil** [comment]
So you're trying to pass `this.props.navigation` into a child component? Have you tried passing it as a prop?

```jsx
render() {
  const { navigation } = this.props;
  return (
    <MyComponent navigation={navigation}/>
  );
}
```

**Comment by puzpuz** [comment]
Yes I've tried that, but `this.props.navigation` is `undefined` inside of `renderRow()` function

FriendListScreen.js
```
  renderRow(friend) {
    return (
      <FriendItem 
        friend={friend} 
        nav={this.props.navigation}
      />
    );
  }
  render() {
    return (
          <List 
            dataArray={friendData}
            renderRow={this.renderRow}
          />
    );
  }
```
FriendItem.js

```
 onDetailPress() {
    this.props.nav.navigate('FriendDetailScreen');
  }
  render() {
    return (
      <View>
              <TouchableOpacity
                  onPress={this.onDetailPress.bind(this)}              
              >
                 <Text>Detail</Text>
              </TouchableOpacity>
      </View>
    );
  }
```

I got this error

![screenshot_1489633499](https://cloud.githubusercontent.com/assets/3850502/23980094/85150af8-0a38-11e7-8aa0-909bac1edcaa.png)


After several readings, I believe the best practice solution for this problem is by using Redux. But I'm still confuse on how to combine Redux with React-Navigation. I'm not sure if this is the best place to ask for that, but I would appreciate if you are willing to help me :)

**Comment by satya164** [comment]
`renderRow = friend => {...}` or `renderRow.bind(this)` should fix it.

**Comment by jmuziki** [comment]
I seem to be having the same problem, but keep getting a "Can't find variable: renderRow: error, code is below:

    renderRow(project) {
        return (
            <ListItem>
                <Text>{project.name}</Text>
            </ListItem>
        );
    }

    return (
        <View>
            <ListView dataArray={projects}
                renderRow={renderRow.bind(this)}
            />
        </View>
    )

**Comment by matthamil** [comment]
@jmuziki If this is inside a class component, it looks like your `renderRow` function is accessible from `this`.

Update the `renderRow` on the `ListView` to this:
`this.renderRow.bind(this)`

Alternatively, you should check out React Native's new components for lists: https://facebook.github.io/react-native/blog/2017/03/13/better-list-views.html

**Comment by jmuziki** [comment]
Thanks @matthamil, that did the trick. When updating my code to do the following:

    renderRow(project) {
        return (
            <ProjectCard project={project}/>
        );
    }

I am unable to access this.props.navigation from within the ProjectCard component. Its undefined within the child

**Comment by matthamil** [comment]
Yes, you would be unable to access it because `navigation` is a prop on the parent component. You should pass the prop down like any other React prop:

```jsx
return (
  <ProjectCard project={project} navigation={this.props.navigation} />
)
```

**Comment by williamdes** [comment]
tl;dr
Works for me :
index.android.js
```jsx
import { StackNavigator } from 'react-navigation';
import Login from './views/Login';
import MainView from './views/MainView';

const MyApp = StackNavigator({
  MainView: { screen: MainView },
  Login: { screen: Login },
  
},
{
  initialRouteName: 'MainView',
  headerMode: ''
});
export default MyApp;
AppRegistry.registerComponent('MyApp', () => MyApp);
```
./views/MainView
```jsx
import React from 'react';
import { Button,Container, Content } from 'native-base';

export default class MainView extends React.Component {
  static navigationOptions = {
    title: 'MainView',
  };
  render() {
  	const { navigate } = this.props.navigation;
    return (
      <Container>
        <Content>
          <Button
          onPress={() => navigate('Login')}
          title=">>Login"
        />
        </Content>
      </Container>
    );
  }
}
```

**Comment by hiteshsahu** [comment]
From Parent screen pass navigator to your row component like a property  

` <MovieRow navigation={this.props.navigation} />
`
And then in List Row component , assign navigate using the props you have passed

` const { navigate } = this.props.navigation;`

Lastly, use `navigate `  variable for screen transition

       ```
   onPress={() => navigate('Detail',
              { Song : 'Love Me',
                  Artist: 'Lucy',
                  Album : "LOVE" ,
                  AlbumArt : "https://cdn.traileraddict.com/content/paramount-pictures/ghost-in-the-shell-poster-8.jpg"
                }
              )}
```

Your code should look like this:
```

export default  class MovieRow extends Component {

  constructor(props) {
     super(props);
  }

  render({onPress} = this.props) {
  const { navigate } = this.props.navigation;
    return (

      <TouchableOpacity
            onPress={() => navigate('Detail',
              { Song : 'Love Me',
                  Artist: 'Lucy',
                  Album : "LOVE" ,
                  AlbumArt : "https://cdn.traileraddict.com/content/paramount-pictures/ghost-in-the-shell-poster-8.jpg"
                }
              )}
              activeOpacity={0.7}
      >
```

For complete source code 

D:\Git Clone Directory\React-Native-Music-Player\src

**Comment by cankatm** [comment]
First import withNavigation:
`import { withNavigation } from 'react-navigation';`

Use:
`export default withNavigation(YourChildComponent);`

If you're using redux, you can do this:
`export default withNavigation(connect(mapStateToProps, actions)(YourChildComponent));`
